espnet2/enh/layers/iva.py

```diff
@@ -0,0 +1,513 @@
+    ----------
+    X: Tensor, shape (..., n_channels, n_frequencies, n_frames)
+        STFT representation of the signal
+    n_iter: int, optional
+        The number of iterations (default 20)
+    n_src: int, optional
+        When n_src is less than the number of channels, the n_src most energetic
+        are selected for the output
+    model: SourceModel
+        The model of source distribution (default: Laplace)
+    eps: float
+        A small constant to make divisions and the like numerically stable
+    two_chan_ip2: bool
+        For the 2 channel case, use the more efficient IP2 algorithm (default: True).
+        Ignored when using more than 2 channels.
+    proj_back_mic: int
+        The microphone index to use as a reference when adjusting the scale and delay
+    use_dmc: bool, optional
+        If True, use checkpointing of the demixing matrix to save memory
+    checkpoints_iter: List of int
+        Optionally, we can keep intermediate results for later analysis
+        Should be used together with checkpoints_list
+    checkpoints_list: List
+        An empty list can be passed and the intermediate signals are
+        appended to the list for iterations numbers contained in checkpoints_iter
+
+    Returns
+    -------
+    X: Tensor, shape (..., n_channels, n_frequencies, n_frames)
+        STFT representation of the signal after separation
+    """
+
+    n_chan, n_freq, n_frames = X.shape[-3:]
+
+    if model is None:
+        model = LaplaceModel()
+
+    # for now, only supports determined case
+    assert callable(model)
+
+    # keep track of original input signal
+    X_original = X.clone()
+
+    if proj_back_mic is not None or use_dmc:
+        assert (
+            0 <= proj_back_mic < n_chan
+        ), "The reference microphone index must be between 0 and # channels - 1."
+        W = X.new_zeros(X.shape[:-3] + (n_freq, n_chan, n_chan))
+        W[:] = torch.eye(n_chan).type_as(W)
+        A = W.clone()
+    else:
+        W = None
+        A = None
+
+    for epoch in range(n_iter):
+
+        if checkpoints_iter is not None and epoch in checkpoints_iter:
+            if use_dmc:
+                X = demix(W, X_original)
+            checkpoints_list.append(X)
+
+        if use_dmc:
+            # use demixing matrix checkpointing
+            model_params = [p for p in self.model.parameters()]
+            W, A = torch.utils.checkpoint.checkpoint(
+                auxiva_iss_one_step_dmc,
+                W,
+                A,
+                model,
+                eps,
+                two_chan_ip2,
+                X_original,
+                *model_params,
+                preserve_rng_state=True
+            )
+        else:
+            # use vanilla backprop
+            X, W, A = auxiva_iss_one_step(X, W, A, model, eps, two_chan_ip2, X_original)
+
+    if use_dmc:
+        X = demix(W, X_original)
+
+    if proj_back_mic is not None:
+        a = A[..., :, [proj_back_mic], :].moveaxis(-1, -3)
```

Comment:
For better compatibility with older PyTorch versions, it it better to replace `.moveaxis()` with `.permute()`.